iroh-relay/src/dns.rs

```diff
@@ -1,22 +1,259 @@
+pub const N0_DNS_NODE_ORIGIN_PROD: &str = "dns.iroh.link";
+/// The n0 testing DNS node origin, for testing.
+pub const N0_DNS_NODE_ORIGIN_STAGING: &str = "staging-dns.iroh.link";
 
-static DNS_RESOLVER: OnceLock<TokioResolver> = OnceLock::new();
+/// The DNS resolver used throughout `iroh`.
+#[derive(Debug, Clone)]
+pub struct DnsResolver(TokioResolver);
 
-/// Get a reference to the default DNS resolver.
-///
-/// The default resolver can be cheaply cloned and is shared throughout the running process.
-/// It is configured to use the system's DNS configuration.
-pub fn default_resolver() -> &'static DnsResolver {
-    DNS_RESOLVER.get_or_init(|| create_default_resolver().expect("unable to create DNS resolver"))
+impl DnsResolver {
+    /// Create a new DNS resolver with sensible cross-platform defaults.
+    ///
+    /// We first try to read the system's resolver from `/etc/resolv.conf`.
+    /// This does not work at least on some Androids, therefore we fallback
+    /// to the default `ResolverConfig` which uses eg. to google's `8.8.8.8` or `8.8.4.4`.
+    pub fn new_with_defaults() -> Self {
+        let (system_config, mut options) =
+            hickory_resolver::system_conf::read_system_conf().unwrap_or_default();
+
+        // Copy all of the system config, but strip the bad windows nameservers.  Unfortunately
+        // there is no easy way to do this.
+        let mut config = hickory_resolver::config::ResolverConfig::new();
+        if let Some(name) = system_config.domain() {
+            config.set_domain(name.clone());
+        }
+        for name in system_config.search() {
+            config.add_search(name.clone());
+        }
+        for nameserver_cfg in system_config.name_servers() {
+            if !WINDOWS_BAD_SITE_LOCAL_DNS_SERVERS.contains(&nameserver_cfg.socket_addr.ip()) {
+                config.add_name_server(nameserver_cfg.clone());
+            }
+        }
+
+        // see [`DnsResolver::lookup_ipv4_ipv6`] for info on why we avoid `LookupIpStrategy::Ipv4AndIpv6`
+        options.ip_strategy = hickory_resolver::config::LookupIpStrategy::Ipv4thenIpv6;
+
+        let resolver = Resolver::tokio(config, options);
+        DnsResolver(resolver)
+    }
+
+    /// Create a new DNS resolver configured with a single UDP DNS nameserver.
+    pub fn new_with_single_nameserver(nameserver: SocketAddr) -> Self {
+        let mut config = hickory_resolver::config::ResolverConfig::new();
+        let nameserver_config = hickory_resolver::config::NameServerConfig::new(
+            nameserver,
+            hickory_resolver::proto::xfer::Protocol::Udp,
+        );
+        config.add_name_server(nameserver_config);
+        DnsResolver(Resolver::tokio(config, Default::default()))
+    }
+
+    /// Removes all entries from the cache.
+    pub fn clear_cache(&self) {
+        self.0.clear_cache();
+    }
+
+    /// Lookup a TXT record.
+    pub async fn lookup_txt(&self, query: impl IntoName) -> Result<TxtLookup, ResolveError> {
```

Comment:
Howe come this doesn't have a timeout while the ipv4 & ipv6 looksups do?

I'm also confused by the return type, it's owned by hickory but not `pub use`?  I'm sure I missed something because otherwise that shouldn't work.

Anyway, long way to say that we should own all our types and not re-export new types from external crates.  We can newtype them.

Reply:
All these methods were so far left as-is from the existing implementations in the extension traits.

I added a timeout to be in line with the others.

Also tried out adding newtypes for `TxtLookup`. It turned out OK. At the same time I though I realized that we still have quite a few `hickory_resolver` or `hickory_proto` types exposed, e.g. the `IntoName` trait is used in most of the methods on `DnsResolver`. Do we want to newtype everything we expose? I'll do a review of the API next, but many of these hickory types are composed or match with other hickory types, so to cover the full API surface of a Dns resolver that will be quite some newtyping, methods etc.

Reply:
The remaining usage of `hickory-proto` is now the `IntoName` trait, which efficiently converts `String`, `&str` and `Name`s into the `Name` struct from `hickory_proto`.